src/Architectures.jl

```diff
@@ -25,23 +26,38 @@ variable `JULIA_NUM_THREADS` is set.
 struct CPU <: AbstractArchitecture end
 
 """
-    GPU <: AbstractArchitecture
+    CUDAGPU <: AbstractArchitecture
 
-Run Oceananigans on a single NVIDIA CUDA GPU.
+Run Oceananigans on a single Nvidia CUDA GPU.
 """
-struct GPU <: AbstractArchitecture end
+struct CUDAGPU <: AbstractArchitecture end
```

Comment:
On another PR (see this file: https://github.com/CliMA/Oceananigans.jl/blob/b7acbf56643bb8e76d8dfc51da407504aebdc76c/src/Architectures.jl) we came up with a slightly different design. We keep `GPU`, but give it a device like this:

```julia
struct GPU{D}
    device :: D
end
```

This is an easy change because we can also introduce the aliases

```julia
const CUDAGPU = GPU{<:CUDAKernels.CUDADevice}
const ROCMGPU = GPU{<:ROCKernels.ROCDevice}
```

which allow most of the subsequent code to remain as is, I think.

The advantage is that users can then write simply `GPU()`, and we will attempt to detect which GPU the user is trying to use. We can also design an API that throws an error or warning if more than one type of device is detected.

A major advantage is this design means that adding AMD support doesn't have to be a breaking change.

Reply:
Love it. I've done this in the latest commit.

Reply:
Hmm.. Trying to set architecture in `example/baroclinic_adjustment.jl` to `GPU()` throws the following 


```
julia --project=. ./bench/baroclinic_adjustment.jl 
ERROR: LoadError: MethodError: no method matching GPU()

Closest candidates are:
  GPU(::D) where D
   @ Oceananigans ~/Oceananigans.jl/src/Architectures.jl:34
```

Reply:
Sure, you added a parameter to the struct and you didn't provide it here.
If you want GPU() to default to something, e.g., GPU() = GPU(DefaultParameter) then you need to define an alias for it.

Reply:
From 

https://github.com/FluidNumerics/Oceananigans.jl/blob/7738002051775eac019d103c34a2d14ead31e86a/src/Architectures.jl#L37-L41

I understand that you need to provide `GPU(CUDA.CUDABackend())` or `GPU(AMDGPU.ROCBackend())` or their shortcuts `CUDAGPU()` or `ROCmGPU()`.

Reply:
I think it's because you removed the `<: AbstractArchitecture` from the GPU struct definition. The constructor expects an AbstractArchitecture and you are providing it with a ROCmGPU which you have not said anywhere that is an AbstractArchitecture

Reply:
Extension of AbstractArchitecture was needed ... got that resolved.